### PR TITLE
merge 'simple-jobs' and 'static-dashboard' into 'master'

### PR DESCRIPTION
**Pull request by jsnal**
Merge all changes from 'simple-job' and 'static-dashboard'

Change log:

- Static front page with HTML cards updating on an interval
- The cards are dynamically created from the nodes in the cluster
- Statistics are formatted as percents out of 100
- Simple job submission API
- Send formatted text from a HTML textarea to the Flask webserver
- Save text to a file within a directory picked by the user
- Run file as a SLURM batch job

### CODE REVIEW

File: src/env.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import os
+
+CLUSTER_SHARED_FS = os.getenv('CLUSTER_SHARED_FS') if os.getenv('CLUSTER_SHARED_FS') else f"/clusterfs/"
+CLUSTER_LOG_PATH  = os.getenv('CLUSTER_LOG_PATH')  if os.getenv('CLUSTER_LOG_PATH')  else f"{CLUSTER_SHARED_FS}/cluster_logs"
+
+
+def _get_init_job_id():
```
Comment (jsnal) [inline src/env.py]: This gets the ID by looking at the sub directories and grabbing the highest number. For example if the directories are 1, 2, 3, 4; the current ID would be 4.

File: src/slurm/sbatch.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+from src.env import get_job_id
+import os
+
+BATCH_OPTS = {
+
+}
+
+def sbatch_run(filepath, *args, **kwargs):
+    batch_command = f'sbatch --output={CLUSTER_LOG_PATH}/{get_job_id()}/{get_job_id()}.out {filepath}'
+    os.system(batch_command)
```
Comment (jsnal) [inline src/slurm/sbatch.py]: This is still a work in progress. Soon we will be adding the ability to change all of the different options within the jobs. Things like node count, core count, timeouts, etc.

File: src/utils/crlf_to_lf.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+def crlf_to_lf(file_path):
+    WINDOWS_LINE_ENDING = b'\r\n'
+    UNIX_LINE_ENDING = b'\n'
+
+    with open(file_path, 'rb') as open_file:
+        content = open_file.read()
+        content = content.replace(WINDOWS_LINE_ENDING, UNIX_LINE_ENDING)
+
+    with open(file_path, 'wb') as open_file:
+        open_file.write(content)
```
Comment (jsnal) [inline src/utils/crlf_to_lf.py]: Replaces windows '\r\n' generated by the HTML with UNIX '\n'.

File: src/env.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    try:
+        return int(max(next(os.walk(CLUSTER_LOG_PATH))[1]))
+    except:
```
Comment (paulcsmith) [inline src/env.py]: I would recommend catching a specific error https://stackoverflow.com/questions/13531247/python-catching-specific-exception

The reason for that is that if at some point you have a legit bug on line 9 and you don't catch a specific error then you may not even realize there's a bug or where the problem is coming for. Usually the more specific the better when catching errors.

File: src/job.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    full_file_path = f'{CLUSTER_LOG_PATH}/{get_job_id()}/{get_job_id()}.{file_ext}'
+
+    try:
+        # Create the job directory and file
+        create_job(posted_file, file_ext, full_file_path)
+
+        # Run the new job as a batch job
+        sbatch_run(full_file_path)
+
+    except Exception:
```
Comment (paulcsmith) [inline src/job.py]: Same here with catching a specific error (or error message). You may also want touse the error message  in the json dump to make it easier to debug from the client side

Comment (paulcsmith) [reply src/job.py]: Though if you really do want to just catch everything a nd return an error JSON object then you can do that, but I'd recommend printing the  exception message and maybe including it in the JSON so it doesn't get totally swallowed

File: src/slurm/sbatch.py (hunk 2 above)
Comment (paulcsmith) [reply src/slurm/sbatch.py]: 👍

File: src/utils/async_request.py [hunk 6]
```diff
@@ -4,9 +4,13 @@
...
     try:
-        async with session.get(url) as response:
+        async with session.get(url, timeout=.5) as response:
             return await response.text()
-    except:
+    except asyncio.TimeoutError:
+        print('Request Timed Out:', url)
+        return None
+    except Exception as e:
+        print('Exception in Request:', url, e)
```
Comment (paulcsmith) [inline src/utils/async_request.py]: Nice! That should help with debugging 👍

File: src/utils/crlf_to_lf.py (hunk 3 above)
Comment (paulcsmith) [reply src/utils/crlf_to_lf.py]: I'd consider adding a comment in the code or extracting a function called `convert_windows_line_endings_to_unix` so the function itself is self documenting

Comment (paulcsmith) [reply src/utils/crlf_to_lf.py]: Actually could you use https://github.com/CamJW101/rpi-cluster-controller/pull/2/files#diff-e59babf30702c5551b2534e7a489e22fR18? If not that's fine, just making sure

**Comment by jsnal** [comment]
Changed some of the things you mentioned in your review. I wasn't exactly sure what you meant by the error catching in src/job.py. I just returned the error message in the json string and printed it to the console so you can see where the code failed.

I also renamed crlf_to_lf to convert_windows_line_endings_to_unix like you recommended.

File: src/job.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        print('Failed creating the job directory:', e)
+        return json.dumps({'status': 'failed creating job', 'code': e})
+
+
+    # Run the new job as a batch job
+    try:
+        sbatch_run(full_file_path)
+    except Exception as e:
+        print('Failed running the batch job:', e)
+        return json.dumps({'status': 'failed running job', 'code': e})
```
Comment (paulcsmith) [inline src/job.py]: Looks good!

**Pull request merged**